test/puppetlabs/puppetdb/testutils/log.clj

```diff
@@ -0,0 +1,164 @@
+(ns puppetlabs.puppetdb.testutils.log
+  (:require
+   [puppetlabs.kitchensink.core :as kitchensink]
+   [puppetlabs.puppetdb.testutils :refer [temp-file]]
+   [me.raynes.fs :as fs])
+  (:import
+   [ch.qos.logback.core Appender FileAppender]
+   [ch.qos.logback.core.spi LifeCycle]
+   [ch.qos.logback.classic Level Logger]
+   [ch.qos.logback.classic.encoder PatternLayoutEncoder]
+   [org.slf4j LoggerFactory]))
+
+(defn- find-logger [id]
+  (.getLogger (LoggerFactory/getILoggerFactory)
+              (if (class? id) id (str id))))
+
+(defmacro with-log-level
+  "Sets the (logback) log level for the logger specified by logger-id
+  during the execution of body.  If logger-id is not a class, it is
+  converted via str, and the level must be a clojure.tools.logging
+  key, i.e. :info, :error, etc."
+  [logger-id level & body]
+  ;; Specify the root logger via org.slf4j.Logger/ROOT_LOGGER_NAME.
+  ;; Assumes use of logback (i.e. logger supports Levels).
+  `(let [logger# (#'puppetlabs.puppetdb.testutils.log/find-logger ~logger-id)
+         original-level# (.getLevel logger#)]
+     (try
+       (.setLevel logger# (case ~level
+                            :trace Level/TRACE
+                            :debug Level/DEBUG
+                            :info Level/INFO
+                            :warn Level/WARN
+                            :error Level/ERROR
+                            :fatal Level/ERROR))
+       (do ~@body)
+       (finally (.setLevel logger# original-level#)))))
+
+(defn- log-event-listener
+  "Returns a log Appender that will call (listen event) for each log event."
+  [listen]
+  ;; No clue yet if we're supposed to start with a default name.
+  (let [name (atom (str "log-listener-" (kitchensink/uuid)))]
+    (reify
+      Appender
+      (doAppend [this event] (listen event))
+      (getName [this] @name)
+      (setName [this x] (reset! name x))
+      LifeCycle
+      (start [this] true)
+      (stop [this] true))))
+
+(defn- with-additional-log-appenders-fn [logger-id appenders body]
+  (let [logger (find-logger logger-id)]
+    (try
+      (doseq [appender appenders]
+        (.addAppender logger appender))
+      (body)
+      (finally
+        (doseq [appender appenders]
```

Comment:
it's just a question of what the implementation of addAppender /
removeAppender actually does. Adding it if it doesn't matter will imply to
future people that it does matter, which I don't want to do.

On Tue, Aug 11, 2015 at 12:08 PM, Rob Browning <notifications@github.com>
wrote:

> In test/puppetlabs/puppetdb/testutils/log.clj
> <https://github.com/puppetlabs/puppetdb/pull/1568#discussion_r36785484>:
>
> > +      Appender
> > +      (doAppend [this event] (listen event))
> > +      (getName [this] @name)
> > +      (setName [this x] (reset! name x))
> > +      LifeCycle
> > +      (start [this] true)
> > +      (stop [this] true))))
> > +
> > +(defn- with-additional-log-appenders-fn [logger-id appenders body]
> > +  (let [logger (find-logger logger-id)]
> > +    (try
> > +      (doseq [appender appenders]
> > +        (.addAppender logger appender))
> > +      (body)
> > +      (finally
> > +        (doseq [appender appenders]
>
> No idea if it matters, but I can reverse it "just in case"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetdb/pull/1568/files#r36785484>.
>